File: android-app/app/src/main/java/com/tbwk/android/TbwkEditor.kt
package com.tbwk.android

import org.w3c.dom.Element
import java.io.ByteArrayInputStream
import java.io.StringWriter
import javax.xml.parsers.DocumentBuilderFactory
import javax.xml.transform.OutputKeys
import javax.xml.transform.TransformerFactory
import javax.xml.transform.dom.DOMSource
import javax.xml.transform.stream.StreamResult

class TbwkEditableDocument private constructor(
    private var file: RawTbwkFile,
) {
    companion object {
        fun parse(bytes: ByteArray): TbwkEditableDocument {
            return TbwkEditableDocument(RawTbwkFile(bytes))
        }
    }

    fun worksheet(): Worksheet = TbwkParser.parse(serializedBytes())

    fun renameMeasurement(index: Int, newName: String) {
        val sanitizedName = newName.trim()
        require(sanitizedName.isNotEmpty()) { "Sample name cannot be empty." }

        val measurementPositions = measurementBlockPositions()
        require(index in measurementPositions.indices) { "Measurement index $index is out of range." }
        file.blocks[measurementPositions[index]].renameMeasurement(sanitizedName)
    }

    fun deleteMeasurement(index: Int) {
        val measurementPositions = measurementBlockPositions()
        require(index in measurementPositions.indices) { "Measurement index $index is out of range." }
        file.blocks.removeAt(measurementPositions[index])
    }

    fun serializedBytes(): ByteArray = file.serializedBytes()

    private fun measurementBlockPositions(): List<Int> {
        return file.blocks.mapIndexedNotNull { position, block ->
            if (block.type == 151L) position else null
        }
    }
}

private class RawTbwkFile(bytes: ByteArray) {
    val header: ByteArray
    val blocks: MutableList<RawTbwkBlock>

    init {
        if (!bytes.startsWithMagicEditor()) {
            throw TbwkParseException("Input is not a TBWK file.")
        }

        val headerSize = bytes.littleEndianUInt32(32).toInt()
        val headerLength = 40 + headerSize
        header = bytes.sliceSafely(0, headerLength)

        val parsedBlocks = mutableListOf<RawTbwkBlock>()
        var offset = headerLength
        while (offset < bytes.size) {
            val block = RawTbwkBlock(bytes, offset)
            parsedBlocks += block
            offset += block.serializedLength()
        }
        blocks = parsedBlocks
    }

    fun serializedBytes(): ByteArray {
        val output = ByteArrayBuilder()
        output.append(header)
        blocks.forEach { output.append(it.serializedBytes()) }
        return output.toByteArray()
    }
}

private class RawTbwkBlock(bytes: ByteArray, offset: Int) {
    var type: Long = bytes.littleEndianUInt32(offset)
    private var blockHeader: ByteArray = bytes.sliceSafely(offset, offset + 12)
    private var leafContent: ByteArray? = null
    private var nestedPrefix: ByteArray? = null
    private var nestedFile: RawTbwkFile? = null

    init {
        val size = bytes.littleEndianUInt32(offset + 4).toInt()
        val content = bytes.sliceSafely(offset + 12, offset + 12 + size)
        if (type in setOf(151L, 920L, 930L) && content.size >= 12) {
            val nestedBytes = content.sliceSafely(12, content.size)
            if (nestedBytes.startsWithMagicEditor()) {
                nestedPrefix = content.sliceSafely(0, 12)
                nestedFile = RawTbwkFile(nestedBytes)
            } else {
                leafContent = content
            }
        } else {
            leafContent = content
        }
    }

    fun renameMeasurement(newName: String) {
        visitBlocks { block ->
            when (block.type) {
                152L -> block.leafContent = renameMeasurementInfo(block.requireLeafContent(), newName)
                931L -> block.leafContent = renameSpectrumMetadata(block.requireLeafContent(), newName)
                62L -> block.leafContent = renameMeasurementXml(block.requireLeafContent(), newName)
            }
        }
    }

    fun serializedLength(): Int = 12 + serializedContent().size

    fun serializedBytes(): ByteArray {
        val content = serializedContent()
        val header = blockHeader.copyOf()
        header.writeUInt32LE(0, type)
        header.writeUInt32LE(4, content.size.toLong())
        return header + content
    }

    private fun serializedContent(): ByteArray {
        val nested = nestedFile
        return if (nested != null) {
            (nestedPrefix ?: byteArrayOf()) + nested.serializedBytes()
        } else {
            leafContent ?: byteArrayOf()
        }
    }

    private fun requireLeafContent(): ByteArray {
        return leafContent ?: throw TbwkParseException("TBWK block $type is not a leaf block.")
    }

    private fun visitBlocks(visitor: (RawTbwkBlock) -> Unit) {
        visitor(this)
        nestedFile?.blocks?.forEach { it.visitBlocks(visitor) }
    }
}

private fun renameMeasurementInfo(content: ByteArray, newName: String): ByteArray {
    val inner = content.sliceSafely(12, content.size)
    val first = inner.tbwkString(0)
    val secondOffset = first.length + 1 + 8
    val second = inner.tbwkString(secondOffset)
    val builder = ByteArrayBuilder()
    builder.append(content.sliceSafely(0, 12))
    builder.append(inner.sliceSafely(0, secondOffset))
    builder.append(tbwkString(newName))
    builder.append(inner.sliceSafely(secondOffset + 1 + second.length, inner.size))
    return builder.toByteArray()
}

private fun renameSpectrumMetadata(content: ByteArray, newName: String): ByteArray {
    var offset = 12
    val first = content.tbwkString(offset)
    offset += 1 + first.length
    val second = content.tbwkString(offset)
    offset += 1 + second.length
    val thirdOffset = offset
    val third = content.tbwkString(thirdOffset)

    val builder = ByteArrayBuilder()
    builder.append(content.sliceSafely(0, thirdOffset))
    builder.append(tbwkString(newName))
    builder.append(content.sliceSafely(thirdOffset + 1 + third.length, content.size))
    return builder.toByteArray()
}

private fun renameMeasurementXml(content: ByteArray, newName: String): ByteArray {
    val xmlBytes = content.sliceSafely(12, content.size)
    val builder = DocumentBuilderFactory.newInstance().newDocumentBuilder()
    val document = builder.parse(ByteArrayInputStream(xmlBytes))
    val sampleNodes = document.getElementsByTagName("VAR")

    var updated = false
    for (index in 0 until sampleNodes.length) {
        val node = sampleNodes.item(index)
        val element = node as? Element ?: continue
        if (element.getAttribute("NAME") == "m_SampleID") {
            element.textContent = newName
            updated = true
        }
    }

    if (!updated) {
        return content
    }

    val transformer = TransformerFactory.newInstance().newTransformer().apply {
        setOutputProperty(OutputKeys.OMIT_XML_DECLARATION, "no")
        setOutputProperty(OutputKeys.ENCODING, "utf-8")
    }
    val writer = StringWriter()
    transformer.transform(DOMSource(document), StreamResult(writer))
    val rebuiltXml = writer.toString().toByteArray(Charsets.UTF_8)
    return content.sliceSafely(0, 12) + rebuiltXml
}

private fun tbwkString(value: String): ByteArray {
    val encoded = value.toByteArray(Charsets.UTF_8)
    require(encoded.size < 256) { "TBWK strings must fit in a single-byte length prefix." }
    return byteArrayOf(encoded.size.toByte()) + encoded
}

private class ByteArrayBuilder {
    private val chunks = ArrayList<ByteArray>()
    private var size = 0

    fun append(bytes: ByteArray) {
        chunks += bytes
        size += bytes.size
    }

    fun toByteArray(): ByteArray {
        val result = ByteArray(size)
        var offset = 0
        chunks.forEach { chunk ->
            chunk.copyInto(result, destinationOffset = offset)
            offset += chunk.size
        }
        return result
    }
}

private fun ByteArray.writeUInt32LE(offset: Int, value: Long) {
    this[offset] = (value and 0xFF).toByte()
    this[offset + 1] = ((value shr 8) and 0xFF).toByte()
    this[offset + 2] = ((value shr 16) and 0xFF).toByte()
    this[offset + 3] = ((value shr 24) and 0xFF).toByte()
}

private fun ByteArray.startsWithMagicEditor(): Boolean {
    if (size < 4) return false
    return this[0] == 0xFE.toByte() &&
        this[1] == 0xFF.toByte() &&
        this[2] == 0xFF.toByte() &&
        this[3] == 0xFF.toByte()
}

private fun ByteArray.sliceSafely(start: Int, endExclusive: Int): ByteArray {
    val lower = kotlin.math.max(0, kotlin.math.min(size, start))
    val upper = kotlin.math.max(lower, kotlin.math.min(size, endExclusive))
    return copyOfRange(lower, upper)
}

private fun ByteArray.littleEndianUInt32(offset: Int): Long {
    var result = 0L
    repeat(4) { index ->
        val actual = offset + index
        if (actual < size) {
            result = result or ((this[actual].toLong() and 0xFF) shl (index * 8))
        }
    }
    return result
}

private data class EditorTbwkString(val length: Int, val value: String)

private fun ByteArray.tbwkString(offset: Int): EditorTbwkString {
    if (offset >= size) throw TbwkParseException("Unexpected end of TBWK string.")
    val length = this[offset].toInt() and 0xFF
    val slice = sliceSafely(offset + 1, offset + 1 + length)
    val value = slice.toString(Charsets.UTF_8)
    return EditorTbwkString(length, value)
}

File: windows-app/src/NanodropViewer.Core/TbwkEditor.cs
using System;
using System.Collections.Generic;
using System.Buffers.Binary;
using System.IO;
using System.Linq;
using System.Text;
using System.Xml.Linq;

namespace NanodropViewer.Core;

public sealed class TbwkEditableDocument
{
    private RawTbwkFile _file;

    private TbwkEditableDocument(RawTbwkFile file)
    {
        _file = file;
    }

    public static TbwkEditableDocument Parse(byte[] bytes)
    {
        return new TbwkEditableDocument(new RawTbwkFile(bytes));
    }

    public Worksheet Worksheet()
    {
        return TbwkParser.Parse(SerializedBytes());
    }

    public void RenameMeasurement(int index, string newName)
    {
        var sanitized = (newName ?? string.Empty).Trim();
        if (string.IsNullOrWhiteSpace(sanitized))
        {
            throw new TbwkException("Sample name cannot be empty.");
        }

        var positions = MeasurementBlockPositions();
        if (index < 0 || index >= positions.Count)
        {
            throw new TbwkException($"Measurement index {index} is out of range.");
        }

        _file.Blocks[positions[index]].RenameMeasurement(sanitized);
    }

    public void DeleteMeasurement(int index)
    {
        var positions = MeasurementBlockPositions();
        if (index < 0 || index >= positions.Count)
        {
            throw new TbwkException($"Measurement index {index} is out of range.");
        }

        _file.Blocks.RemoveAt(positions[index]);
    }

    public byte[] SerializedBytes()
    {
        return _file.SerializedBytes();
    }

    public void Save(string path)
    {
        File.WriteAllBytes(path, SerializedBytes());
    }

    private List<int> MeasurementBlockPositions()
    {
        return _file.Blocks
            .Select((block, position) => new { block.Type, Position = position })
            .Where(item => item.Type == 151L)
            .Select(item => item.Position)
            .ToList();
    }

    private sealed class RawTbwkFile
    {
        public RawTbwkFile(byte[] bytes)
        {
            ReadOnlySpan<byte> span = bytes;
            if (!span.StartsWithTbwkMagic())
            {
                throw new TbwkException("Input is not a TBWK file.");
            }

            var headerSize = checked((int)span.ReadUInt32LittleEndian(32));
            var headerLength = 40 + headerSize;
            Header = bytes.AsSpan(0, headerLength).ToArray();

            Blocks = new List<RawTbwkBlock>();
            var offset = headerLength;
            while (offset < bytes.Length)
            {
                var block = new RawTbwkBlock(bytes, offset);
                Blocks.Add(block);
                offset += block.SerializedLength();
            }
        }

        public byte[] Header { get; }
        public List<RawTbwkBlock> Blocks { get; }

        public byte[] SerializedBytes()
        {
            var chunks = new List<byte[]> { Header };
            var totalSize = Header.Length;

            foreach (var block in Blocks)
            {
                var bytes = block.SerializedBytes();
                chunks.Add(bytes);
                totalSize += bytes.Length;
            }

            var result = new byte[totalSize];
            var offset = 0;
            foreach (var chunk in chunks)
            {
                Buffer.BlockCopy(chunk, 0, result, offset, chunk.Length);
                offset += chunk.Length;
            }

            return result;
        }
    }

    private sealed class RawTbwkBlock
    {
        private readonly byte[] _blockHeader;
        private byte[]? _leafContent;
        private byte[]? _nestedPrefix;
        private RawTbwkFile? _nestedFile;

        public RawTbwkBlock(byte[] bytes, int offset)
        {
            ReadOnlySpan<byte> span = bytes;
            Type = span.ReadUInt32LittleEndian(offset);
            _blockHeader = bytes.AsSpan(offset, 12).ToArray();

            var size = checked((int)span.ReadUInt32LittleEndian(offset + 4));
            var content = bytes.AsSpan(offset + 12, size).ToArray();
            if ((Type == 151L || Type == 920L || Type == 930L) && content.Length >= 12)
            {
                var nestedBytes = content.AsSpan(12).ToArray();
                if (((ReadOnlySpan<byte>)nestedBytes).StartsWithTbwkMagic())
                {
                    _nestedPrefix = content.AsSpan(0, 12).ToArray();
                    _nestedFile = new RawTbwkFile(nestedBytes);
                }
                else
                {
                    _leafContent = content;
                }
            }
            else
            {
                _leafContent = content;
            }
        }

        public long Type { get; private set; }

        public void RenameMeasurement(string newName)
        {
            VisitBlocks(block =>
            {
                switch (block.Type)
                {
                    case 152L:
                        block._leafContent = RenameMeasurementInfo(block.RequireLeafContent(), newName);
                        break;
                    case 931L:
                        block._leafContent = RenameSpectrumMetadata(block.RequireLeafContent(), newName);
                        break;
                    case 62L:
                        block._leafContent = RenameMeasurementXml(block.RequireLeafContent(), newName);
                        break;
                }
            });
        }

        public int SerializedLength()
        {
            return 12 + SerializedContent().Length;
        }

        public byte[] SerializedBytes()
        {
            var content = SerializedContent();
            var header = _blockHeader.ToArray();
            BinaryPrimitives.WriteUInt32LittleEndian(header.AsSpan(0, 4), (uint)Type);
            BinaryPrimitives.WriteUInt32LittleEndian(header.AsSpan(4, 4), (uint)content.Length);
            return header.Concat(content).ToArray();
        }

        private byte[] SerializedContent()
        {
            if (_nestedFile is not null)
            {
                return (_nestedPrefix ?? Array.Empty<byte>()).Concat(_nestedFile.SerializedBytes()).ToArray();
            }

            return _leafContent ?? Array.Empty<byte>();
        }

        private byte[] RequireLeafContent()
        {
            return _leafContent ?? throw new TbwkException($"TBWK block {Type} is not a leaf block.");
        }

        private void VisitBlocks(Action<RawTbwkBlock> visitor)
        {
            visitor(this);
            if (_nestedFile is null)
            {
                return;
            }

            foreach (var block in _nestedFile.Blocks)
            {
                block.VisitBlocks(visitor);
            }
        }
    }

    private static byte[] RenameMeasurementInfo(byte[] content, string newName)
    {
        var inner = content.AsSpan(12).ToArray();
        var first = inner.AsSpan().ReadTbwkString(0);
        var secondOffset = first.Length + 1 + 8;
        var second = inner.AsSpan().ReadTbwkString(secondOffset);

        return content.AsSpan(0, 12).ToArray()
            .Concat(inner.AsSpan(0, secondOffset).ToArray())
            .Concat(TbwkString(newName))
            .Concat(inner.AsSpan(secondOffset + 1 + second.Length).ToArray())
            .ToArray();
    }

    private static byte[] RenameSpectrumMetadata(byte[] content, string newName)
    {
        var offset = 12;
        var first = content.AsSpan().ReadTbwkString(offset);
        offset += 1 + first.Length;
        var second = content.AsSpan().ReadTbwkString(offset);
        offset += 1 + second.Length;
        var thirdOffset = offset;
        var third = content.AsSpan().ReadTbwkString(thirdOffset);

        return content.AsSpan(0, thirdOffset).ToArray()
            .Concat(TbwkString(newName))
            .Concat(content.AsSpan(thirdOffset + 1 + third.Length).ToArray())
            .ToArray();
    }

    private static byte[] RenameMeasurementXml(byte[] content, string newName)
    {
        var xmlBytes = content.AsSpan(12).ToArray();
        var root = XDocument.Parse(Encoding.UTF8.GetString(xmlBytes)).Root;
        if (root is null)
        {
            return content;
        }

        var updated = false;
        foreach (var element in root.Descendants("VAR"))
        {
            if (string.Equals(element.Attribute("NAME")?.Value, "m_SampleID", StringComparison.Ordinal))
            {
                element.Value = newName;
                updated = true;
            }
        }

        if (!updated)
        {
            return content;
        }

        var rebuiltXml = Encoding.UTF8.GetBytes(root.Document?.ToString(SaveOptions.DisableFormatting) ?? root.ToString(SaveOptions.DisableFormatting));
        return content.AsSpan(0, 12).ToArray().Concat(rebuiltXml).ToArray();
    }

    private static byte[] TbwkString(string value)
    {
        var encoded = Encoding.UTF8.GetBytes(value);
        if (encoded.Length >= 256)
        {
            throw new TbwkException("TBWK strings must fit in a single-byte length prefix.");
        }

        return new[] { (byte)encoded.Length }.Concat(encoded).ToArray();
    }
}
